firmware/bb_frc_controller_pololu_32u4_robot_controller/bb_frc_controller_pololu_32u4_robot_controller.ino

```diff
@@ -1,19 +1,377 @@
+            }
+            break;
+        case INPUT:
+            if (BRD_IS_DIGITAL_PIN(pin)) {
+                pinMode(digitalPinMap[pin], INPUT); // disable output driver
+                Firmata.setPinMode(pin, INPUT);
+            }
+            break;
+        case PIN_MODE_PULLUP:
+            if (BRD_IS_DIGITAL_PIN(pin)) {
+                pinMode(digitalPinMap[pin], INPUT_PULLUP);
+                Firmata.setPinMode(pin, PIN_MODE_PULLUP);
+                Firmata.setPinState(pin, 0);
+            }
+            break;
+        case OUTPUT:
+            if (BRD_IS_DIGITAL_PIN(pin)) {
+                pinMode(digitalPinMap[pin], OUTPUT);
+                Firmata.setPinMode(pin, OUTPUT);
+            }
+        default:
+            Firmata.sendString("Unsupported pin mode");
+    }
+}
+
+void setPinValueCallback(byte pin, int value) {
+    if (BRD_IS_DIGITAL_PIN(pin)) {
+        if (Firmata.getPinMode(pin) == OUTPUT) {
+            Firmata.setPinState(pin, value);
+            digitalWrite(digitalPinMap[pin], value);
+        }
+    }
+}
+
+void analogWriteCallback(byte pin, int value) {
+    if (pin < BRD_NUM_SERVO_PINS) {
+        // Special case for pin 0 and 1, which map to motors
+        if (pin < 2) {
+            // TODO motor control here
+
+        }
+        else {
+            pin -= 2; // drop the pin count to get zero indexed servos
+            servos[pin].write(value);
+            Firmata.setPinState(servoPinMap[pin+2], value);
+        }
+    }
+}
+
+void digitalWriteCallback(byte port, int value) {
+    // We get sent a port (8 bits) to write to. This does NOT map to the
+    // physical port. We'll probably ALWAYS get port 0, and value is a bitmask
+    // of what to write
+
+    for (byte pin = 0; pin < BRD_NUM_DIGITAL_PINS; pin++) {
+        byte val = (value >> pin) & 0x1;
+        setPinValueCallback(pin, val == 0 ? LOW : HIGH);
+    }
+}
+
+void reportAnalogCallback(byte analogPin, int value) {
+    if (analogPin < BRD_NUM_ANALOG_PINS) {
+        if (value == 0) {
+            analogInputsToReport = analogInputsToReport & ~ (1 << analogPin);
+        }
+        else {
+            analogInputsToReport = analogInputsToReport | (1 << analogPin);
+
+            // prevent during system reset, or all analog pins will be reported
+            // which might report noise for unconnected pins
+            if (!isResetting) {
+                // Send pin value immediately
+                Firmata.sendAnalog(analogPin, analogRead(analogPin));
+            }
+        }
+    }
+}
+
+void reportDigitalCallback(byte port, int value) {
+    // We are only getting port 0, and value is a bitmask of which pins to read
+    reportPins = (byte)value;
+
+    if (value) {
+        // Send the port value right away
+        outputDigitalPort(readDigitalPort(), true);
+    }
+}
+
+void systemResetCallback() {
+    isResetting = true;
+
+    // Reset all digital pins to be output only
+    // Detach all servo pins
```

Comment:
Where is the code to detach the servos and disable the motors?

Reply:
Haven’t implemented that part yet. We might not need to detach servos actually

Reply:
Fair, All other code not implemented has a TODO. Would be good to add here to just to keep track.